system/jlib/jptree.cpp

```diff
@@ -531,6 +543,377 @@ const char *queryHead(const char *xpath, StringBuffer &head)
+    {
+    }
+    ~CQualifierMap()
+    {
+        for (auto &e: attrValueMaps)
+            delete e.second;
+    }
+    CValueMap *addMapping(const char *lhs, IPTArrayValue &array)
+    {
+        CValueMap *valueMap = new CValueMap(lhs, array);
+        attrValueMaps.emplace(std::make_pair(std::string(lhs), valueMap));
+        return valueMap;
+    }
+    CValueMap *addMappingIfNew(const char *lhs, IPTArrayValue &array)
+    {
+        CriticalBlock b(crit);
+        auto it = attrValueMaps.find(lhs);
+        if (attrValueMaps.find(lhs) == attrValueMaps.end())
+            return addMapping(lhs, array);
+        else
+            return it->second;
+    }
+    void addMatchingValues(const IPropertyTree *tree)
+    {
+        for (auto &e: attrValueMaps)
+        {
+            const char *v = tree->queryProp(e.first.c_str());
+            if (v)
+                e.second->insertEntry(v, tree);
+        }
+    }
+    void removeMatchingValues(const IPropertyTree *tree)
+    {
+        for (auto &e: attrValueMaps)
+        {
+            const char *lhsp = e.first.c_str();
+            const char *oldV = tree->queryProp(lhsp);
+            if (oldV)
+                verifyex(e.second->removeEntry(oldV, tree));
+        }
+    }
+    void replaceMatchingValues(const IPropertyTree *oldTree, const IPropertyTree *newTree)
+    {
+        for (auto &e: attrValueMaps)
+        {
+            const char *lhsp = e.first.c_str();
+            const char *oldV = oldTree->queryProp(lhsp);
+            if (oldV)
+            {
+                verifyex(e.second->removeEntry(oldV, oldTree));
+                const char *newV = newTree->queryProp(lhsp);
+                if (newV)
+                    e.second->insertEntry(newV, newTree);
+            }
+        }
+    }
+    CValueMap *find(const char *lhs)
+    {
+        auto it = attrValueMaps.find(lhs);
+        if (it == attrValueMaps.end())
+            return nullptr;
+        return it->second;
+    }
+    void removeEntryIfMapped(const char *lhs, const char *v, const IPropertyTree *tree)
+    {
+        auto it = attrValueMaps.find(lhs);
+        if (it != attrValueMaps.end())
+            it->second->removeEntry(v, tree);
+    }
+    void insertEntryIfMapped(const char *lhs, const char *v, const IPropertyTree *tree)
+    {
+        auto it = attrValueMaps.find(lhs);
+        if (it != attrValueMaps.end())
+            it->second->insertEntry(v, tree);
+    }
+    void replaceEntryIfMapped(const char *lhs, const char *oldv, const char *newv, const IPropertyTree *tree)
+    {
+        auto it = attrValueMaps.find(lhs);
+        if (it != attrValueMaps.end())
+            it->second->replaceEntry(oldv, newv, tree);
+    }
+};
+
+// parse qualifier, returns true if simple equality expression found
+static bool parseEqualityQualifier(const char *&xxpath, unsigned &lhsLen, const char *&rhsBegin, unsigned &rhsLen)
+{
+    const char *xpath = xxpath;
+    while (*xpath == ' ' || *xpath == '\t') xpath++;
```

Comment:
Is xpath always ok, as in -
while (xpath && (*xpath == ' ' || *xpath == '\t')) xpath++;

Reply:
>Is xpath always ok

yes, it is guaranteed to be non-null, called from checkMapIterator(), called from getElements().

Reply:
ok, thx